### PR TITLE
linux: use preferred_gcc instead of gcc

### PR DESCRIPTION
**Pull request by iMichka**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

### CODE REVIEW

**Comment by iMichka** [comment]
I need to run some sanity checks locally before shipping this. Proposed by @sjackman

File: Library/Homebrew/extend/os/linux/keg_relocate.rb [hunk 1]
```diff
@@ -84,9 +86,9 @@ def self.relocation_formulae
   def self.bottle_dependencies
     @bottle_dependencies ||= begin
       formulae = relocation_formulae
-      gcc = Formula["gcc"]
+      gcc = Formula[preferred_gcc]
```
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/keg_relocate.rb:89]:
```suggestion
      gcc = Formulary.factory(preferred_gcc)
```
… for consistency with the rest of the Homebrew/brew code base.

File: Library/Homebrew/extend/ENV/shared.rb [hunk 2]
```diff
@@ -314,7 +314,7 @@ def gcc_version_formula(name)
     version = name[GNU_GCC_REGEXP, 1]
     gcc_version_name = "gcc@#{version}"
 
-    gcc = Formulary.factory("gcc")
+    gcc = Formulary.factory(preferred_gcc)
```
Comment (sjackman) [inline Library/Homebrew/extend/ENV/shared.rb:317]:
```suggestion
    gcc = Formulary.factory("gcc")
```
This one should stay as is. It translates `gcc-10` to the formula `gcc`, which is correct on both macOS and Linux.

File: Library/Homebrew/extend/os/linux/keg_relocate.rb [hunk 3]
```diff
@@ -84,9 +86,9 @@ def self.relocation_formulae
...
       if !Homebrew::EnvConfig.force_homebrew_on_linux? &&
-         DevelopmentTools.non_apple_gcc_version("gcc") < gcc.version.to_i
+         DevelopmentTools.non_apple_gcc_version(preferred_gcc) < gcc.version.to_i
```
Comment (sjackman) [inline Library/Homebrew/extend/os/linux/keg_relocate.rb:91]:
```suggestion
         DevelopmentTools.non_apple_gcc_version("gcc") < gcc.version.to_i
```
This one needs to stay as is, it tests the version of whichever `gcc --version` is found first in the `PATH`. Line 87 immediately above however needs to change from `gcc = Formula["gcc"]` to `gcc = Formulary.factory("preferred_gcc")`

**Review by sjackman — CHANGES_REQUESTED** [review]

**Comment by sjackman** [comment]
In `extend/os/linux/linkage_checker.rb`
```diff
-@undeclared_deps -= ["gcc", "glibc"]
+@undeclared_deps -= [preferred_compiler, "glibc"]
```

**Comment by sjackman** [comment]
In `extend/os/linux/system_config.rb`
```diff
-["glibc", "gcc", "xorg"].each do |f|
+["glibc", preferred_compiler, "xorg"].each do |f|
```

**Review by MikeMcQuaid — APPROVED** [review]
👍🏻 once this is 💚. I really like how minimal and clean this change is ❤️

File: Library/Homebrew/compilers.rb [hunk 4]
```diff
@@ -110,13 +110,13 @@ def compiler
...
-  def preferred_gcc
+  def self.preferred_gcc
     "gcc"
   end
 
+  private
+
   def gnu_gcc_versions
     # prioritize gcc version provided by gcc formula.
     v = Formulary.factory(preferred_gcc).version.to_s.slice(/\d+/)
```
Comment (sjackman) [inline Library/Homebrew/compilers.rb:122]:
```suggestion
    v = Formulary.factory(CompilerSelector.preferred_gcc).version.to_s.slice(/\d+/)
```

**Comment by iMichka** [comment]
I had to stub gcc@5 in the tests to make it pass. That's the only test change I had to make.

**Review by sjackman — APPROVED** [review]

**Comment by sjackman** [comment]
I did some quick tests in the `homebrew/brew` Docker image.
✅ `brew install -sv hello` uses `/usr/bin/gcc-9` provided by the host
✅ `brew install gcc@5 && brew install -sv hello` uses `/home/linuxbrew/.linuxbrew/bin/gcc-5`

Currently `gcc` and `gcc@5` conflict, because they both provide GCC 5 until https://github.com/Homebrew/linuxbrew-core/pull/21380 is merged.

I need to think about migration issues for existing users that have `gcc` installed, because their host system is old (GCC < 5). We want it to smoothly install `gcc@5` for those users (and upgrade `gcc` to 10, but I'm less concerned about that). I don't believe `gcc@5` will be installed automatically, but I need to test that.

**Comment by sjackman** [comment]
> We want it to smoothly install gcc@5 for those users (and upgrade gcc to 10, but I'm less concerned about that). I don't believe gcc@5 will be installed automatically, but I need to test that.

This sort of works! Once `gcc` is upgraded to GCC 10 and no longer provides `~/.linuxbrew/opt/gcc/bin/gcc`, when the user has no other executable named `gcc` version ≥ 5 in their `PATH`, then Homebrew will install `preferred_gcc`, aka `gcc@5`.

**Comment by iMichka** [comment]
Regarding the transition phase from gcc to gcc-5 in linuxbrew-core: stuff might get broken for complex cases, especially for people which do not have gcc-5 installed. But I think we can guide them to run some commands manually to clean the mess up if anything goes wrong. But your tests tend to demonstrate that there will be no issue.

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by sjackman** [comment]
Thank you, Michka!

**Pull request merged**